### ISSUE TITLE
How to use `lag()` in the relational API?

### ISSUE THREAD

**Issue by krlmlr** [duckdb/duckdb#6946]
### What happens?

This works in the CLI:

```
D create table x (a int);
D insert into x values (1), (2);
D select * from x;
┌───────┐
│   a   │
│ int32 │
├───────┤
│     1 │
│     2 │
└───────┘
D select a, lag(a) over () from x;
┌───────┬────────────────┐
│   a   │ lag(a) OVER () │
│ int32 │     int32      │
├───────┼────────────────┤
│     1 │                │
│     2 │              1 │
└───────┴────────────────┘
```

How to implement the same in the relational API, perhaps also with non-empty `OVER ()` ?

### To Reproduce

``` r
con <- DBI::dbConnect(duckdb::duckdb())
experimental <- FALSE
df1 <- data.frame(a = 1:6, g = c(3L, 3L, 3L, 4L, 4L, 4L))

rel1 <- duckdb:::rel_from_df(con, df1, experimental = experimental)
rel2 <- duckdb:::rel_project(
  rel1,
  list(
    {
      tmp_expr <- duckdb:::expr_reference("a")
      duckdb:::expr_set_alias(tmp_expr, "a")
      tmp_expr
    },
    {
      tmp_expr <- duckdb:::expr_reference("g")
      duckdb:::expr_set_alias(tmp_expr, "g")
      tmp_expr
    },
    {
      tmp_expr <- duckdb:::expr_window(
        duckdb:::expr_function("lag", list(duckdb:::expr_reference("a"))),
        list(duckdb:::expr_reference("g"))
      )
      duckdb:::expr_set_alias(tmp_expr, "d")
      tmp_expr
    }
  )
)
rel2
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a, g as g, lag(a, NULL, NULL) OVER (PARTITION BY g) as d]
#>   r_dataframe_scan(0x10a866108)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (INTEGER)
#> - g (INTEGER)
#> - d (INTEGER)
duckdb:::rel_to_altrep(rel2)
#>   a g d
#> 1 1 3 1
#> 2 2 3 2
#> 3 3 3 3
#> 4 4 4 4
#> 5 5 4 5
#> 6 6 4 6
```

<sup>Created on 2023-04-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### OS:

macOS aarch64

### DuckDB Version:

3dcba6d0b59d50a5cb23a9878ed2847f654712d4

### DuckDB Client:

R

### Full Name:

Kirill Müller

### Affiliation:

cynkra GmbH

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by hawkfish** [comment]
I thought this was implemented with PR #6757?

**Comment by krlmlr** [comment]
Thanks. It does seem we need `offset_expr` as a constant:

``` r
con <- DBI::dbConnect(duckdb::duckdb())
experimental <- FALSE
df1 <- data.frame(a = 1:6, g = c(3L, 3L, 3L, 4L, 4L, 4L))

rel1 <- duckdb:::rel_from_df(con, df1, experimental = experimental)
rel2 <- duckdb:::rel_project(
  rel1,
  list(
    {
      tmp_expr <- duckdb:::expr_reference("a")
      duckdb:::expr_set_alias(tmp_expr, "a")
      tmp_expr
    },
    {
      tmp_expr <- duckdb:::expr_reference("g")
      duckdb:::expr_set_alias(tmp_expr, "g")
      tmp_expr
    },
    {
      tmp_expr <- duckdb:::expr_window(
        duckdb:::expr_function("lag", list(duckdb:::expr_reference("a"))),
        list(duckdb:::expr_reference("g")),
        offset_expr=duckdb:::expr_constant(1)
      )
      duckdb:::expr_set_alias(tmp_expr, "d")
      tmp_expr
    }
  )
)

rel2
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [a as a, g as g, lag(a, 1.0, NULL) OVER (PARTITION BY g) as d]
#>   r_dataframe_scan(0x1332ed008)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - a (INTEGER)
#> - g (INTEGER)
#> - d (INTEGER)

duckdb:::rel_to_altrep(rel2)
#>   a g  d
#> 1 1 3 NA
#> 2 2 3  1
#> 3 3 3  2
#> 4 4 4 NA
#> 5 5 4  4
#> 6 6 4  5
```

<sup>Created on 2023-04-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
